src/meepgeom.cpp

```diff
@@ -977,50 +987,96 @@ static bool get_front_object(const meep::volume &v, geom_box_tree geometry_tree,
              ((id >= id1 && id >= id2) && (id1 == id2 || material_type_equal(mat1, mat2)))) {
       o2 = o;
       shiftby2 = shiftby;
       id2 = id;
       mat2 = mat;
+      p2 = q;
     }
     else if (!(id1 < id2 && (id1 == id || material_type_equal(mat1, mat))) &&
-             !(id2 < id1 && (id2 == id || material_type_equal(mat2, mat))))
+             !(id2 < id1 && (id2 == id || material_type_equal(mat2, mat)))){
       return false;
+      }
   }
 
   // CHECK(id1 > -1, "bug in object_of_point_in_tree?");
   if (id2 == -1) { /* only one nearby object/material */
     id2 = id1;
     o2 = o1;
     mat2 = mat1;
+    p2 = p1;
     shiftby2 = shiftby1;
   }
 
-  if ((o1 && is_variable(o1->material)) || (o2 && is_variable(o2->material)) ||
-      ((is_variable(default_material) || is_file(default_material)) &&
-       (!o1 || is_file(o1->material) || !o2 || is_file(o2->material))))
-    return false;
-
   if (id1 >= id2) {
     *o_front = o1;
     shiftby_front = shiftby1;
     mat_front = mat1;
-    if (id1 == id2)
+    p_front = p1;
+    if (id1 == id2) {
       mat_behind = mat1;
-    else
+      p_behind = p1;
+    }
+    else {
       mat_behind = mat2;
+      p_behind = p2;
+    }
   }
   if (id2 > id1) {
     *o_front = o2;
     shiftby_front = shiftby2;
     mat_front = mat2;
+    p_front = p2;
     mat_behind = mat1;
+    p_behind = p1;
   }
   return true;
 }
 
+double get_material_grid_fill(meep::ndim dim, double d, double r, double u, double eta,
+  bool *mg_averaging, bool *center_is_void){
+  /* Lets assume that the "background material" is void (u=0) and that the
+  "foreground material is solid (u=1)". The fill fraction then describes
+  the amount of foreground material inhabits the cell.
+
+  The analytic volume expressions for the end cap only specify the fill fraction
+  of the cap relative to the whole sphere... but we don't know what the cap 
+  actually is (solid or void). So we need a little extra logic to sort that out.
+  
+  Occasionally, the distance to the nearest interface is outside the current
+  sphere, which means we don't need to do any averaging (the fill is 0 or 1). Again,
+  we don't know if that means we are in void or solid, however, until we look
+  at u. To make things easy, we'll assume the cap is solid until the end, when we
+  can verify.
+  */
+  double rel_fill;
+  if (abs(d) >= abs(r)){
+    *mg_averaging = false;
+    return -1.0; // garbage fill
+  } else {
+    if (dim == meep::D1)
+      rel_fill = (r-d)/(2*r);
+    else if (dim == meep::D2 || dim == meep::Dcyl){
+      rel_fill = (1/(r*r*meep::pi)) * (r*r*std::acos(d/r)-d*std::sqrt(r*r-d*d));
+    }
+    else if (dim == meep::D3)
+      rel_fill = (((r-d)*(r-d))/(4*meep::pi*r*r*r))*(2*r+d);
+  }
+
+  *mg_averaging = true;
+  if (u <= eta){
+    *center_is_void = true;
+    return rel_fill;   // center is void, so cap must be sold
+  }else{
+    *center_is_void = false;
+    return 1-rel_fill; // center is solid, so cap must be void
+  }
```

Comment:
May not be needed any more — just return the fill fraction of the u=1 material always, for example.